File: ai/java/image-classification/src/test/java/com/rbleggi/imageclassification/ImageClassificationTest.java
package com.rbleggi.imageclassification;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.*;

class ImageClassificationTest {
    private List<Image> trainingSet;
    private Image araraImage;
    private Image oncaImage;

    @BeforeEach
    void setUp() {
        trainingSet = List.of(
            new Image("t1", "arara_azul.jpg", new ImageFeatures(0.2, 0.6, 0.8, 0.7, 0.5, 0.8), "Arara"),
            new Image("t2", "arara_vermelha.jpg", new ImageFeatures(0.8, 0.5, 0.3, 0.6, 0.6, 0.7), "Arara"),
            new Image("t3", "onca_pintada.jpg", new ImageFeatures(0.6, 0.5, 0.3, 0.4, 0.8, 0.6), "Onca"),
            new Image("t4", "onca_preta.jpg", new ImageFeatures(0.3, 0.3, 0.2, 0.3, 0.7, 0.4), "Onca"),
            new Image("t5", "capivara.jpg", new ImageFeatures(0.5, 0.4, 0.3, 0.5, 0.4, 0.3), "Capivara")
        );

        araraImage = new Image("test1", "arara_nova.jpg", new ImageFeatures(0.3, 0.7, 0.9, 0.75, 0.6, 0.85), "Arara");
        oncaImage = new Image("test2", "onca_nova.jpg", new ImageFeatures(0.4, 0.4, 0.25, 0.35, 0.75, 0.5), "Onca");
    }

    @Test
    @DisplayName("ThresholdStrategy classifies based on color thresholds")
    void thresholdStrategy_classifiesBasedOnThresholds() {
        ImageClassificationSystem system = new ImageClassificationSystem(new ThresholdStrategy());
        Classification result = system.classify(araraImage, trainingSet);

        assertNotNull(result.predictedCategory());
        assertTrue(result.confidence() >= 0.0 && result.confidence() <= 1.0);
        assertEquals("Threshold-based", result.method());
    }

    @Test
    @DisplayName("KNNStrategy classifies using K nearest neighbors")
    void knnStrategy_usesNearestNeighbors() {
        ImageClassificationSystem system = new ImageClassificationSystem(new KNNStrategy(3));
        Classification result = system.classify(araraImage, trainingSet);

        assertEquals("Arara", result.predictedCategory());
        assertTrue(result.confidence() > 0.0);
        assertTrue(result.method().contains("KNN"));
    }

    @Test
    @DisplayName("KNNStrategy confidence reflects vote proportion")
    void knnStrategy_confidenceReflectsVotes() {
        ImageClassificationSystem system = new ImageClassificationSystem(new KNNStrategy(3));
        Classification result = system.classify(araraImage, trainingSet);

        assertTrue(result.confidence() <= 1.0);
        assertTrue(result.confidence() >= 0.33);
    }

    @Test
    @DisplayName("NeuralNetworkStrategy uses weighted scoring")
    void neuralNetworkStrategy_usesWeightedScoring() {
        ImageClassificationSystem system = new ImageClassificationSystem(new NeuralNetworkStrategy());
        Classification result = system.classify(araraImage, trainingSet);

        assertNotNull(result.predictedCategory());
        assertTrue(result.confidence() > 0.0 && result.confidence() <= 1.0);
        assertEquals("Neural-Network", result.method());
    }

    @Test
    @DisplayName("ImageClassificationSystem processes batch of images")
    void classificationSystem_processesBatch() {
        ImageClassificationSystem system = new ImageClassificationSystem(new KNNStrategy(3));
        List<Image> testSet = List.of(araraImage, oncaImage);
        List<Classification> results = system.classifyBatch(testSet, trainingSet);

        assertEquals(2, results.size());
        assertEquals("Arara", results.get(0).predictedCategory());
    }

    @Test
    @DisplayName("ImageClassificationSystem calculates accuracy")
    void classificationSystem_calculatesAccuracy() {
        ImageClassificationSystem system = new ImageClassificationSystem(new KNNStrategy(3));
        List<Image> testSet = List.of(araraImage, oncaImage);
        double accuracy = system.calculateAccuracy(testSet, trainingSet);

        assertTrue(accuracy >= 0.0 && accuracy <= 1.0);
    }

    @Test
    @DisplayName("Different strategies produce different results")
    void differentStrategies_produceDifferentResults() {
        ImageClassificationSystem threshold = new ImageClassificationSystem(new ThresholdStrategy());
        ImageClassificationSystem knn = new ImageClassificationSystem(new KNNStrategy(3));
        ImageClassificationSystem nn = new ImageClassificationSystem(new NeuralNetworkStrategy());

        Classification r1 = threshold.classify(araraImage, trainingSet);
        Classification r2 = knn.classify(araraImage, trainingSet);
        Classification r3 = nn.classify(araraImage, trainingSet);

        assertNotEquals(r1.method(), r2.method());
        assertNotEquals(r2.method(), r3.method());
    }
}

File: ai/kotlin/image-classification/src/test/kotlin/com/rbleggi/imageclassification/ImageClassificationTest.kt
package com.rbleggi.imageclassification

import kotlin.test.*

class ImageClassificationTest {

    private val trainingSet = listOf(
        Image("t1", "Onca 1", listOf(0.9, 0.8, 0.2, 0.1), Category.Onca),
        Image("t2", "Onca 2", listOf(0.85, 0.75, 0.25, 0.15), Category.Onca),
        Image("t3", "Arara 1", listOf(0.2, 0.3, 0.9, 0.8), Category.Arara),
        Image("t4", "Arara 2", listOf(0.25, 0.35, 0.85, 0.75), Category.Arara),
        Image("t5", "Capivara 1", listOf(0.5, 0.6, 0.4, 0.5), Category.Capivara)
    )

    private val testImage = Image("test", "Test", listOf(0.88, 0.78, 0.22, 0.12), Category.Onca)

    @Test
    fun `ThresholdStrategy classifies image correctly`() {
        val thresholds = mapOf(
            Category.Onca to listOf(0.9, 0.8, 0.2, 0.1),
            Category.Arara to listOf(0.2, 0.3, 0.9, 0.8),
            Category.Capivara to listOf(0.5, 0.6, 0.4, 0.5)
        )
        val strategy = ThresholdStrategy(thresholds)
        val result = strategy.classify(testImage)

        assertEquals(Category.Onca, result.category)
        assertTrue(result.confidence > 0.0)
    }

    @Test
    fun `ThresholdStrategy provides confidence score`() {
        val thresholds = mapOf(
            Category.Onca to listOf(0.9, 0.8, 0.2, 0.1)
        )
        val strategy = ThresholdStrategy(thresholds)
        val result = strategy.classify(testImage)

        assertTrue(result.confidence >= 0.0 && result.confidence <= 1.0)
        assertTrue(result.details.contains("Threshold matching"))
    }

    @Test
    fun `KNNStrategy classifies based on nearest neighbors`() {
        val strategy = KNNStrategy(trainingSet, k = 3)
        val result = strategy.classify(testImage)

        assertEquals(Category.Onca, result.category)
        assertTrue(result.confidence > 0.0)
    }

    @Test
    fun `KNNStrategy respects k parameter`() {
        val strategy = KNNStrategy(trainingSet, k = 1)
        val result = strategy.classify(testImage)

        assertTrue(result.details.contains("k=1"))
        assertTrue(result.confidence >= 0.0 && result.confidence <= 1.0)
    }

    @Test
    fun `KNNStrategy uses euclidean distance`() {
        val oncaImage = Image("onca", "Onca", listOf(0.9, 0.8, 0.2, 0.1), Category.Onca)
        val strategy = KNNStrategy(trainingSet, k = 3)
        val result = strategy.classify(oncaImage)

        assertEquals(Category.Onca, result.category)
    }

    @Test
    fun `NeuralNetworkLikeStrategy classifies with weights`() {
        val weights = mapOf(
            Category.Onca to listOf(0.8, 0.7, -0.3, -0.2),
            Category.Arara to listOf(-0.3, -0.2, 0.8, 0.7),
            Category.Capivara to listOf(0.1, 0.2, 0.1, 0.2)
        )
        val strategy = NeuralNetworkLikeStrategy(weights)
        val result = strategy.classify(testImage)

        assertNotNull(result.category)
        assertTrue(result.confidence > 0.0)
    }

    @Test
    fun `NeuralNetworkLikeStrategy uses sigmoid activation`() {
        val weights = mapOf(
            Category.Onca to listOf(1.0, 1.0, 1.0, 1.0)
        )
        val strategy = NeuralNetworkLikeStrategy(weights)
        val result = strategy.classify(testImage)

        assertTrue(result.confidence > 0.0 && result.confidence < 1.0)
        assertTrue(result.details.contains("Rede neural"))
    }

    @Test
    fun `ImageClassifier classifies single image`() {
        val classifier = ImageClassifier(KNNStrategy(trainingSet, k = 3))
        val result = classifier.classify(testImage)

        assertEquals(Category.Onca, result.category)
    }

    @Test
    fun `ImageClassifier classifies batch of images`() {
        val classifier = ImageClassifier(KNNStrategy(trainingSet, k = 3))
        val images = listOf(testImage, testImage)
        val results = classifier.classifyBatch(images)

        assertEquals(2, results.size)
        assertTrue(results.all { it.category == Category.Onca })
    }

    @Test
    fun `ImageClassifier calculates accuracy`() {
        val classifier = ImageClassifier(KNNStrategy(trainingSet, k = 3))
        val testImages = listOf(
            Image("t1", "Test1", listOf(0.9, 0.8, 0.2, 0.1), Category.Onca),
            Image("t2", "Test2", listOf(0.2, 0.3, 0.9, 0.8), Category.Arara)
        )
        val accuracy = classifier.accuracy(testImages)

        assertTrue(accuracy >= 0.0 && accuracy <= 1.0)
    }

    @Test
    fun `Image data class stores features correctly`() {
        val image = Image("id", "name", listOf(1.0, 2.0, 3.0), Category.Onca)

        assertEquals("id", image.id)
        assertEquals("name", image.name)
        assertEquals(3, image.features.size)
        assertEquals(Category.Onca, image.actualCategory)
    }

    @Test
    fun `ClassificationResult contains required fields`() {
        val result = ClassificationResult(Category.Onca, 0.95, "test details")

        assertEquals(Category.Onca, result.category)
        assertEquals(0.95, result.confidence)
        assertEquals("test details", result.details)
    }

    @Test
    fun `Category enum has all Brazilian options`() {
        assertNotNull(Category.Onca)
        assertNotNull(Category.Arara)
        assertNotNull(Category.Capivara)
        assertNotNull(Category.PaoDeQueijo)
        assertNotNull(Category.Feijoada)
        assertNotNull(Category.Acai)
    }

    @Test
    fun `ThresholdStrategy handles multiple categories`() {
        val thresholds = mapOf(
            Category.Onca to listOf(0.9, 0.8, 0.2, 0.1),
            Category.Arara to listOf(0.2, 0.3, 0.9, 0.8),
            Category.Capivara to listOf(0.5, 0.6, 0.4, 0.5),
            Category.PaoDeQueijo to listOf(0.7, 0.7, 0.7, 0.7)
        )
        val strategy = ThresholdStrategy(thresholds)
        val result = strategy.classify(testImage)

        assertNotNull(result.category)
    }

    @Test
    fun `KNNStrategy with k=1 returns single neighbor category`() {
        val strategy = KNNStrategy(trainingSet, k = 1)
        val result = strategy.classify(testImage)

        assertTrue(result.confidence == 1.0)
    }

    @Test
    fun `NeuralNetworkLikeStrategy handles negative weights`() {
        val weights = mapOf(
            Category.Onca to listOf(-1.0, -1.0, -1.0, -1.0)
        )
        val strategy = NeuralNetworkLikeStrategy(weights)
        val result = strategy.classify(testImage)

        assertTrue(result.confidence > 0.0)
    }

    @Test
    fun `ImageClassifier accuracy is 100 percent for perfect predictions`() {
        val perfectTraining = listOf(testImage)
        val classifier = ImageClassifier(KNNStrategy(perfectTraining, k = 1))
        val accuracy = classifier.accuracy(listOf(testImage))

        assertEquals(1.0, accuracy)
    }
}
